### ISSUE TITLE
JSMin minifier is non-free (suggest JSqueeze or JShrink as alternatives)

### ISSUE THREAD

**Issue by AdamWill** [owncloud/core#13052]
As explained at https://github.com/owncloud/core/pull/11430#issuecomment-68317924 , the JSMin minifier which is used by that PR is non-free according to [FSF](https://www.gnu.org/licenses/license-list.html#JSON), [Debian](https://lists.debian.org/debian-lint-maint/2013/01/msg00153.html), and [Fedora](https://fedoraproject.org/wiki/Licensing:Main#Bad_Licenses). Also, its original upstream developer [abandoned it years ago and explicitly recommends other alternatives](https://github.com/rgrove/jsmin-php/blob/master/README.md). mrclay is maintaining the copy in his repo, but probably not as actively as the newer alternatives.

I [ran some benchmarks](https://www.happyassassin.net/2014/12/29/adventures-in-php-web-asset-minimization/) which indicated JSqueeze might be the best replacement, at least from a numbers perspective. It's also actively maintained by a well-known developer. Its [lack of namespacing](https://github.com/tchwork/jsqueeze/issues/11) is a bit of a pain for distros but in the worst case (if they don't take my suggestion to namespace) can be dealt with.

More importantly, though, I managed to get a hacky build of master up and running and tested it with JSqueeze, and it, er, [doesn't work](https://github.com/tchwork/jsqueeze/issues/14) (for me, anyway, confirmation would be welcome). Upstream seems fairly responsive to bug reports, so I'm hoping they'll get to that one soon.

The other alternative that looked decent to me was [JShrink](https://github.com/tedious/JShrink). It's currently-maintained, namespaced, travis-ified, psr-0 compliant, lots of buzzword boxes. It doesn't perform as well as JSqueeze in terms of speed or file size, but it *does* work with OC for basic use (I just tested), which is a significant advantage.

I'll note an ancillary benefit, here: pulling in mrclay/minify just to get a JavaScript minifier is kind of atomic-weapon-nutcracker stuff. As my blog post (linked earlier) points out, it's a sort of poor man's Assetic with eight different CSS/JS minification implementations and a whole bunch of other nutty stuff in it. Switching to JShrink or JSqueeze would allow us to just have a JavaScript minifier, not a JavaScript minifier tied to three other JavaScript minifiers and lord knows what else. As it stands, OC 8 will be shipping (by my count) four CSS minifiers (the three in mrclay/minify, plus the one it actually *uses*), two full PHP JavaScript minifiers, and two different Closure Compiler wrappers/interfaces (and that's not even counting any others lying around in the source tree...)

I can send PRs for switching to either JSqueeze or JShrink pretty easily - the changes to OC itself are trivial, then it's just a 3rdparty commit that drops mrclay/minify and pulls in the one we want - but it seemed prudent to wait for a response from JSqueeze to my bug report.

Edit: oh, for the record, a drawback of JShrink is that Assetic doesn't currently ship a filter for it. There's a [github project which adds one](https://github.com/nibsirahsieu/SalvaJshrinkBundle) - for my test I just dropped that straight into the Assetic tree and it worked fine, so it could be sent upstream pretty easily if they'd take it. But they do have a JSqueeze filter already, so it's one less step there.

**Comment by AdamWill** [comment]
Tested and confirmed fix with JSqueeze 2.0.1, but 2.0.1 also namespaces the lib (at my request, sorry :>) so Assetic's filter doesn't work OOTB any more. I've sent a PR to Assetic for that: https://github.com/kriswallsmith/assetic/pull/698

**Comment by MorrisJobke** [comment]
IBM has a changed version of this:

```

**Comment by MorrisJobke** [comment]
IBM has a changed version of this:

```
IBM, its customers, partners, and minions, to use JSLint for evil.
```

http://dev.hasenj.org/post/3272592502/ibm-and-its-minions

;)

**Comment by MorrisJobke** [comment]
@AdamWill Thanks for spotting this!

**Comment by AdamWill** [comment]
@MorrisJobke that's still not really sufficient as the exception doesn't apply to all (unless you consider everyone a 'minion' of IBM...), so you still effectively can't distribute the software on free terms to all. Everyone who's not a 'customer, partner or minion' of IBM still can't use it for evil. In any case I don't think you could hold that it applies to the copy of `JSMin.php` in minify.

In any case, JSqueeze or JShrink or Matthias' 'minify' would be superior choices for several other reasons - they're currently maintained, follow good current coding practices, are faster and/or more efficient than JSMin, and don't come attached to the rest of the crazy stuff in `mrclay/minify`. Switching just seems like the sensible thing to do. I'm hoping upstream Assetic will pick up my PR soon, then it should be fine to use JSqueeze, I think (though if folks can help test it in various setups that'd obviously help).

**Comment by MorrisJobke** [comment]
@AdamWill Sure

**Comment by AdamWill** [comment]
assetic's now merged my filter change onto master, but not onto the 1.2 branch, yet - I've asked if they can merge it to 1.2.

**Comment by MorrisJobke** [comment]
@karlitschek @DeepDiver1975 @AdamWill As we are in feature freeze what if the upstream project can't provide the update fast enough? Is there an alternative plan?

**Comment by karlitschek** [comment]
@DeepDiver1975 What is your opinion?

**Comment by AdamWill** [comment]
@MorrisJobke I can see three options:

1. carry a downstream tweak of some kind to allow use of JSqueeze (backport the filter file patch to Assetic, or just dump an extra filter file in there, or patch the bug fix into JSqueeze 1.x, or something)

2. Use JShrink - current OC master works with current upstream JShrink fine, for me. This requires changing a couple of lines in `lib/private/templatelayout.php`, pulling JShrink into 3rdparty, and ideally dropping minify from 3rdparty (as it wouldn't be needed any more).

3. Use JSMinPlus, which is already in the current `minify` codebase - all this requires is a change of a couple of lines in `lib/private/templatelayout.php`. I don't know your rules/guidelines about non-free code, but in this case you might want/need to strip the non-free bits from the bundled `minify` (anything JSMin-derived).

**Comment by DeepDiver1975** [comment]
Okay - let's go for JSMinPlus for OC8 because it requires NO updates of 3rdparty libs - which reduces risk and test effort. For OC8.1 we can reevaluate (and maybe even allow some config option for installations who want to run e.g. YUI tools ???)

@AdamWill delivery of non-free code is not that of an issue of owncloud itself - but we want to collaborate with linux distributions as good as possible and respect the guidelines to not ship non-free code. For OC8 these code pieces have to be stripped. I hope this works out for you.

I'll take care about updating to JSMinPlus .....

**Comment by AdamWill** [comment]
shouldn't be a huge deal for downstreams - my approach will probably be simply to package jsminplus on its own (or bundle it with OC since it'll only be used temporarily). less hassle than packaging minify with all the non-free bits stripped out.

**Comment by DeepDiver1975** [comment]
grrr - JSMinPlus generates broken JS

**Comment by AdamWill** [comment]
bah :( I guess I didn't test it, I must've only tested JSqueeze and JShrink.

**Comment by AdamWill** [comment]
So to get *something* helpful done, I drafted up the necessary changes to use JShrink:

https://github.com/owncloud/core/pull/13185
https://github.com/owncloud/3rdparty/pull/149

if it would help I can, as an alternative, draft up the changes to use JSqueeze (which would be along the same lines, but probably with an added Assetic filter called 'JSqueeze2Filter' or so in `lib/private/assetic` , along the lines of the recently-added `SeparatorFilter`).

**Comment by karlitschek** [comment]
@DeepDiver1975 What do you suggest?

**Comment by AdamWill** [comment]
@karlitschek there's been more activity on the PR I submitted to switch to JSqueeze: https://github.com/owncloud/core/pull/13229 . At first we found several cases where the minified JS misbehaved, but most of them seem to be fixed now, and it's close to being mergeable, I think.

**Comment by PVince81** [comment]
Any update on this ? Or defer to 8.2 ?

**Comment by AdamWill** [comment]
sorry, I'm on vacation till Jun 8, I'll get back to it then.

**Comment by cmonteroluque** [comment]
@AdamWill any update? We need to make a quick determination to keep this in 8.2. Thanks!

**Comment by AdamWill** [comment]
This has been done, so far as I'm concerned, for months. It just never gets merged when I update it so every two fucking months I have to jump through the 're-generate the 3rdparty PR' hoop *again*. I'm sick of it.

https://github.com/owncloud/core/pull/13229
https://github.com/owncloud/3rdparty/pull/150

the last time I jumped through the hoops was July 18, where @pvince81 asked some people to review it, and no-one really did (except for one comment on the 3rdparty PR which I explained is just what composer does when you use it as documented; if someone knows an alternative way to do it, i'm all ears).
